pandas/core/indexes/datetimes.py

```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
     # Wrapping DatetimeArray
 
     @property
-    def _eadata(self):
-        return DatetimeArray._simple_new(self._data,
-                                         tz=self.tz, freq=self.freq)
+    def _data(self):
+        return self._eadata._data
+
+    @property
+    def _freq(self):
+        return self._eadata._freq
+
+    @_freq.setter
+    def _freq(self, value):
+        # Validation will be handled by _eadata setter
+        self._eadata._freq = value
+
+    @property
+    def freq(self):
+        return self._freq
+
+    @freq.setter
+    def freq(self, value):
+        # validation is handled by _eadata setter
+        self._eadata.freq = value
+
+    @property
+    def _tz(self):
+        return self._eadata._tz
+
+    @property
+    def tz(self):
+        # GH#18595
+        return self._tz
+
+    @tz.setter
+    def tz(self, value):
+        # GH 3746: Prevent localizing or converting the index by setting tz
+        raise AttributeError("Cannot directly set timezone. Use tz_localize() "
```

Comment:
I wonder if this can just do `self._eadata.tz = value`, which will raise the AttributeError for us? Then we don't duplicate this code.